kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1511 @@
+    					return framework.NewStatus(framework.Error, "pod hostname not found")
+    				}
+    
+    				// Get zone and region from Pod Hostname
+    				regionPodHostname := networkAwareUtil.GetNodeRegion(podHostname.Node())
+    				zonePodHostname:= networkAwareUtil.GetNodeZone(podHostname.Node())
+    
+    				if regionPodHostname == "" && zonePodHostname == "" { // Node has no zone and region defined
+    					numNotOK += 1
+    				} else if region == regionPodHostname{ // If Nodes belong to the same region
+    					if zone == zonePodHostname{ // If Nodes belong to the same zone
+    						numOK += 1
+    					} else { // belong to a different zone, check maxNetworkCost
+    						cost, costOK := costMap[networkAwareUtil.CostKey{ // Retrieve the cost from the map (origin: zone, destination: pod zoneHostname)
+    							Origin:      zone, // Time Complexity: O(1)
+    							Destination: zonePodHostname,
+    						}]
+    						if costOK {
+    							if cost <= d.MaxNetworkCost{
+    								numOK += 1
+    							}else{
+    								numNotOK += 1
+    							}
+    						}
+    					}
+    				} else{ // belong to a different region
+    					cost, costOK := costMap[networkAwareUtil.CostKey{ // Retrieve the cost from the map (origin: zone, destination: pod zoneHostname)
+    						Origin:      region, // Time Complexity: O(1)
+    						Destination: regionPodHostname,
+    					}]
+    					if costOK{
+    						if cost <= d.MaxNetworkCost{
+    							numOK += 1
+    						}else{
+    							numNotOK += 1
+    						}
+    					}
+    				}
+    			}
+    		}
+    	}
+    }
+    klog.V(6).Infof("NumNotOk: %v / numOK: %v ", numNotOK, numOK)
+    
+    if numNotOK > numOK{
+    	return framework.NewStatus(framework.Unschedulable,
+    		fmt.Sprintf("Node %v does not meet several network requirements from Pod dependencies: OK: %v NotOK: %v", nodeInfo.Node().Name, numOK, numNotOK))
+    }
+    return nil
+}
+```
+
+#### `NetworkOverhead` Filter Example
+
+Let's consider the following AppGroup CRD for the appGroup `A1` containing three workloads representing pods `P1 - P3`:
+
+```yaml
+# Example App Group CR spec
+apiVersion: scheduling.sigs.k8s.io/v1alpha1
+kind: AppGroup
+metadata:
+  name: a1
+spec:
+  numMembers: 3
+  topologySortingAlgorithm: KahnSort
+  workloads:
+    - workloadName: P1
+      workloadRef: 
+        kind: Pod 
+        name: P1
+      dependencies:
+        - workloadName: P2
+          minBandwidth: "100Mi"
+          maxNetworkCost: 15
+    - workloadName: P2
+      workloadRef: 
+          kind: Pod
+          name: P2
+      dependencies:
+        - workloadName: P3
+          minBandwidth: "250Mi"
+          maxNetworkCost: 20
+    - workloadName: P3
+      workloadRef:
+        kind: Pod 
+        name: P3
+```
+
+At a given moment, the status part is the following: 
+
+ ```yaml
+# Status from App Group CRD 
+(...)        
+          status: 
+            properties:
+              scheduled:
```

Comment:
It looks weird to persistent the status of scheduled pods and use it as the source-of-truth. Can you use the {wordload}Lister in the scheduler plugin to obtain the wordload's status? (which is covered by that workload controller). In this way, it'd offload the AppGroup controller's responsibility a lot.

Reply:
In fact, we already noticed in our early implementation that since the controller works async it is not fast enough to update the CR accordingly if all pods belonging to an AppGroup are deployed together. So, we modified the plugin implementation to work with an AppGroup lister instead. 

We planned to keep the Status part of the CR just for logging/record purposes. We can remove it completely and keep the controller’s responsibility simpler. 

We updated the plugins descriptions/examples accordingly.